nav2_costmap_2d/src/collision_checker.cpp

```diff
@@ -0,0 +1,209 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_costmap_2d/collision_checker.hpp"
+#include "nav2_costmap_2d/cost_values.hpp"
+#include "nav2_util/line_iterator.hpp"
+#include "nav2_costmap_2d/exceptions.hpp"
+#include "nav2_costmap_2d/footprint.hpp"
+
+namespace nav2_costmap_2d
+{
+
+CollisionChecker::CollisionChecker(
+  rclcpp::Node::SharedPtr ros_node,
+  std::shared_ptr<CostmapSubscriber> costmap_sub,
+  std::shared_ptr<FootprintSubscriber> footprint_sub,
+  tf2_ros::Buffer & tf_buffer,
+  std::string name)
+: tf_buffer_(tf_buffer),
+  node_(ros_node), name_(name),
+  costmap_sub_(costmap_sub), footprint_sub_(footprint_sub)
+{
+  node_->get_parameter_or<std::string>(
+    "global_frame", global_frame_, std::string("map"));
+  node_->get_parameter_or<std::string>(
+    "robot_base_frame", robot_base_frame_, std::string("base_link"));
+}
+
+CollisionChecker::~CollisionChecker() {}
+
+bool CollisionChecker::isCollisionFree(
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  try {
+    if (scorePose(pose) < 0) {
+      return false;
+    }
+    return true;
+  } catch (const IllegalPoseException & e) {
+    RCLCPP_ERROR(node_->get_logger(), "%s", e.what());
+    return false;
+  } catch (const CollisionCheckerException & e) {
+    RCLCPP_ERROR(node_->get_logger(), "%s", e.what());
+    return false;
+  }
+}
+
+double CollisionChecker::scorePose(
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  Costmap2D * costmap_;
+  try {
+    costmap_ = costmap_sub_->getCostmap();
+  } catch (const std::runtime_error & e) {
+    throw CollisionCheckerException(e.what());
+  }
+
+  unsigned int cell_x, cell_y;
+  if (!costmap_->worldToMap(pose.x, pose.y, cell_x, cell_y)) {
+    RCLCPP_DEBUG(node_->get_logger(), "Map Cell: [%d, %d]", cell_x, cell_y);
+    throw IllegalPoseException(name_, "Pose Goes Off Grid.");
+  }
+
+  Footprint footprint;
+  if (!footprint_sub_->getFootprint(footprint)) {
+    throw CollisionCheckerException("Footprint not available.");
+  }
+
+  Footprint footprint_spec;
+  unorientFootprint(footprint, footprint_spec);
+  transformFootprint(pose.x, pose.y, pose.theta, footprint_spec, footprint);
+
+  // now we really have to lay down the footprint in the costmap grid
+  unsigned int x0, x1, y0, y1;
+  double line_cost = 0.0;
+  double footprint_cost = 0.0;
+
+  // we need to rasterize each line in the footprint
+  for (unsigned int i = 0; i < footprint.size() - 1; ++i) {
+    // get the cell coord of the first point
+    if (!costmap_->worldToMap(footprint[i].x, footprint[i].y, x0, y0)) {
```

Comment:
Making these checks a separate function might reduce some duplication `worldToMap(wx, wy, mx, mw)`

Reply:
and maybe have this iteration over all points inside an other function `validateFootprint`.

Reply:
This for loop does more than just validate the footprint, it also gets the map coordinates which are passed through to `lineCost(x0, x1, y0, y1)` for each point pair.